AI/Nullkiller/Engine/PriorityEvaluator.cpp

```diff
@@ -1259,15 +1358,203 @@ float PriorityEvaluator::evaluate(Goals::TSubgoal task)
 
 		engine->process();
 
-		result = value->getValue();
+		fuzzyResult = value->getValue();
 	}
-	catch(fl::Exception & fe)
+	catch (fl::Exception& fe)
 	{
 		logAi->error("evaluate VisitTile: %s", fe.getWhat());
 	}
+	if (ai->settings->isUseFuzzy())
+	{
+		result = fuzzyResult;
+	}
+	else
+	{
+		float score = 0;
+		float maxWillingToLose = ai->cb->getTownsInfo().empty() ? 1 : 0.25;
+
+		logAi->trace("BEFORE: priorityTier %d, Evaluated %s, loss: %f, turn: %d, turns main: %f, scout: %f, gold: %f, cost: %d, army gain: %f, army growth: %f skill: %f danger: %d, threatTurns: %d, threat: %d, role: %s, strategical value: %f, conquest value: %f cwr: %f, fear: %f, isDefend: %d, fuzzy: %f",
+			priorityTier,
+			task->toString(),
+			evaluationContext.armyLossPersentage,
+			(int)evaluationContext.turn,
+			evaluationContext.movementCostByRole[HeroRole::MAIN],
+			evaluationContext.movementCostByRole[HeroRole::SCOUT],
+			goldRewardPerTurn,
+			evaluationContext.goldCost,
+			evaluationContext.armyReward,
+			evaluationContext.armyGrowth,
+			evaluationContext.skillReward,
+			evaluationContext.danger,
+			evaluationContext.threatTurns,
+			evaluationContext.threat,
+			evaluationContext.heroRole == HeroRole::MAIN ? "main" : "scout",
+			evaluationContext.strategicalValue,
+			evaluationContext.conquestValue,
+			evaluationContext.closestWayRatio,
+			evaluationContext.enemyHeroDangerRatio,
+			evaluationContext.isDefend,
+			fuzzyResult);
+
+		switch (priorityTier)
+		{
+			case PriorityTier::INSTAKILL: //Take towns / kill heroes in immediate reach
+			{
+				if (evaluationContext.turn > 0)
+					return 0;
+				if(evaluationContext.conquestValue > 0)
+					score = 1000;
+				if (score == 0 || (evaluationContext.enemyHeroDangerRatio > 1 && (evaluationContext.turn > 0 || evaluationContext.isExchange) && !ai->cb->getTownsInfo().empty()))
```

Comment:
More floating point comparisons here (and below)
Either use integer type for score or vstd::isAlmostZero for comparing with zero

Reply:
Replaced these with vstd::isAlmostZero aswell.